src/main/java/org/wildfly/security/auth/server/RealmIdentity.java

```diff
@@ -266,4 +284,28 @@ public boolean exists() throws RealmUnavailableException {
             return false;
         }
     };
+
+    /**
+     * Represents a key that is specific to the realm and used to uniquely identify a {@link RealmIdentity} instance.
+     */
+    interface Key {
+
+        Key EMPTY = (Key) () -> null;
+
+        /**
+         * Returns the value of this key.
+         *
+         * @return the value of this key
+         */
+        Object getValue();
```

Comment:
If it's just a Object wrapper, there's no value to the Key in the first place.  The key should either be opaque or we should just drop it in favor of Object.  An opaque key should have a specific contract in terms of hashability.